### PR TITLE
Add support for records

### LINKED ISSUE #7: Add support for records => Rust tuples
We (tandemdrive) have to wrangle some existing queries into a rotate to column first, then zip in order to avoid using records.

We can avoid significant complexity by implementing support for converting PostgreSQL records into Rust tuples.

### PR DESCRIPTION
**Pull request by remkop22**
Closes #7

### CODE REVIEW

File: tokio-postgres/tests/test/main.rs [hunk 1]
```diff
@@ -997,3 +997,27 @@ async fn query_scalar_opt() {
 
     assert_eq!(age, None);
 }
+
+#[tokio::test]
+async fn records() {
```
Comment (rikvdkleij) [inline tokio-postgres/tests/test/main.rs:1002]: Maybe split in 2 tests?

File: postgres-types/src/lib.rs [hunk 2]
```diff
@@ -775,12 +775,67 @@ impl<'a> FromSql<'a> for SystemTime {
...
+                    let mut fields = record.fields();
+
+                    Ok((
+                        $({
+                            let field = fields.next()?
+                                .ok_or_else(|| -> Box<dyn Error + Sync + Send> {
+                                    "not enough fields for record".into()
+                                })?;
+
+                            let ty = Type::from_oid(field.field_type())
```
Comment (rikvdkleij) [inline postgres-types/src/lib.rs:808]: This can fail at runtime if type is custom type?

File: postgres-types/src/lib.rs [hunk 3]
```diff
@@ -775,12 +775,67 @@ impl<'a> FromSql<'a> for SystemTime {
...
+                                    "could not convert type for record field".into()
+                                })?;
+
+                            $Type::from_sql_nullable(&ty, field.bytes())?
+                        }),*
+                    ))
+                }
+
+                fn accepts(ty: &Type) -> bool {
+                    *ty == Type::RECORD && *ty.kind() == Kind::Pseudo
```
Comment (rikvdkleij) [inline postgres-types/src/lib.rs:819]: Why also accept a `Pseudo` type kind?

File: postgres-protocol/src/types/mod.rs [hunk 4]
```diff
@@ -657,6 +657,108 @@ impl<'a> FallibleIterator for ArrayValues<'a> {
...
+                return Err("invalid value length".into());
+            }
+
+            let (bytes, buf) = self.buf.split_at(len);
+            self.buf = buf;
+
+            Ok(Some(RecordField {
+                field_type,
+                bytes: Some(bytes),
+            }))
```
Comment (lucasholten-ds) [inline postgres-protocol/src/types/mod.rs:737]: Maybe de-duplicate this? only the `bytes` field is different between branches.

File: postgres-protocol/src/types/mod.rs [hunk 5]
```diff
@@ -657,6 +657,108 @@ impl<'a> FallibleIterator for ArrayValues<'a> {
...
+
+            return Ok(None);
+        }
+
+        self.remaining -= 1;
+
+        let field_type = self.buf.read_u32::<BigEndian>()?;
+        let len = self.buf.read_i32::<BigEndian>()?;
+
+        if len < 0 {
```
Comment (lucasholten-ds) [inline postgres-protocol/src/types/mod.rs:719]: Can the length be negative?

Comment (remkop22) [reply postgres-protocol/src/types/mod.rs:719]: Yeah postgres makes a distinction between for a example an empty string, which would have 0 bytes and `null` which would be negative.

File: postgres-types/src/lib.rs (hunk 3 above)
Comment (remkop22) [reply postgres-types/src/lib.rs:819]: This is more of a `tokio-postgres` concept. It has `Type` and a type can have a `Kind`. 

If you return a column with a predefined postgres composite type it returns a `Type(Inner::Other(Arc(Other)))`. this type returns `Kind::Composite(_)` when you call `Type::kind()` which contains all the fields of the composite type already resolved.

If you now return a column with an anonymous record type, it returns a `Type(Inner::Record)` and if you call `Type::kind()` it returns `Kind::Pseudo`. 

So to make the distinction between named composites and anonymous records we check if it is both a `Inner::Record` and a `Kind::Pseudo`

File: postgres-types/src/lib.rs (hunk 2 above)
Comment (remkop22) [reply postgres-types/src/lib.rs:808]: Yes it can indeed

File: postgres-protocol/src/types/mod.rs (hunk 4 above)
Comment (remkop22) [reply postgres-protocol/src/types/mod.rs]: Done!

File: tokio-postgres/tests/test/main.rs (hunk 1 above)
Comment (remkop22) [reply tokio-postgres/tests/test/main.rs:1002]: Done!

File: postgres-types/src/lib.rs (hunk 2 above)
Comment (rikvdkleij) [reply postgres-types/src/lib.rs:808]: Ok, maybe add warning (or in the error message) that only built-in types are supported at the moment. It's very common in TD to use types like `td_code` and `td_numeric_12`

Comment (rikvdkleij) [reply postgres-types/src/lib.rs:808]: I mean, hopefully we can support also non-built-in types in the future which can be done in a separate PR.

File: postgres-types/src/lib.rs (hunk 3 above)
Comment (rikvdkleij) [reply postgres-types/src/lib.rs:819]: Ok, I see

**Review by rikvdkleij — APPROVED** [review]
:+1:

**Pull request merged**